src/client/wire_protocol/operations.rs

```diff
@@ -0,0 +1,309 @@
+
+        let bson_length = match length_of_bson(&query) {
+            Ok(i) => i,
+            Err(_) => return Err("Unable to serialize query".to_owned())
+        };
+
+        // Add the length of the optional BSON document only if it exists.
+        let option_length = match return_field_selector {
+            Some(ref bson) => match length_of_bson(bson) {
+                Ok(i) => i,
+                Err(_) => return Err("Unable to serialize return_field_selector".to_owned())
+            },
+            None => 0
+        };
+
+        let total_length = header_length + i32_length + string_length +
+                           bson_length + option_length;
+
+        let header = Header::new_query(total_length, header_request_id);
+
+        Ok(Message::OpQuery { header: header, flags: flags,
+                              full_collection_name: full_collection_name,
+                              number_to_skip: number_to_skip,
+                              number_to_return: number_to_return, query: query,
+                              return_field_selector: return_field_selector })
+    }
+
+    /// Writes a serialized BSON document to a given buffer.
+    ///
+    /// # Arguments
+    ///
+    /// `buffer` - The buffer to write to.
+    /// `bson` - The document to serialize and write.
+    ///
+    /// # Return value
+    ///
+    /// Returns nothing on success, or an error string on failure.
+    fn write_bson_document(buffer: &mut Write, bson: &BSON) -> Result<(), String>{
+        let mut temp_buffer = vec![];
+
+        match bson::encode_document(&mut temp_buffer, bson) {
+            Ok(_) => match buffer.write(&temp_buffer) {
+                Ok(_) => Ok(()),
+                Err(_) => Err("unable to write BSON".to_owned())
+            },
+            Err(_) => Err("unable to encode bson".to_owned())
+        }
+    }
+
+    /// Writes a serialized query message to a given buffer.
+    ///
+    /// # Arguments
+    ///
+    /// `buffer` - The buffer to write to.
+    /// `header` - The header for the given message.
+    /// `flags` - Bit vector of query option.
+    /// `full_collection_name` - The full qualified name of the collection,
+    ///                          beginning with the database name and a dot.
+    /// `number_to_skip` - The number of initial documents to skip over in the
+    ///                    query results.
+    /// `number_to_return - The total number of documents that should be
+    ///                     returned by the query.
+    /// `return_field_selector - An optional projection of which fields should
+    ///                          be present in the documents to be returned by
+    ///                          the query.
+    ///
+    /// # Return value
+    ///
+    /// Returns nothing on success, or an error string on failure.
+    fn write_query(buffer: &mut Write, header: &Header,
+                   flags: i32, full_collection_name: &str,
```

Comment:
it might make for easier comprehension if flags are broken out into a boolean struct instead of just a i32 value delegating the encoding to a higher level of the driver.